packages/eslint-plugin/src/rules/no-invalid-this.ts

```diff
@@ -0,0 +1,86 @@
+import { TSESTree } from '@typescript-eslint/experimental-utils';
+import baseRule from 'eslint/lib/rules/no-invalid-this';
+import {
+  InferOptionsTypeFromRule,
+  InferMessageIdsTypeFromRule,
+  createRule,
+  deepMerge,
+} from '../util';
+
+export type Options = InferOptionsTypeFromRule<typeof baseRule>;
+export type MessageIds = InferMessageIdsTypeFromRule<typeof baseRule>;
+
+const schema = deepMerge(
+  Array.isArray(baseRule.meta.schema)
+    ? baseRule.meta.schema[0]
+    : baseRule.meta.schema,
+  {
+    properties: {
+      capIsConstructor: {
+        type: 'boolean',
+        default: true,
+      },
+    },
+  },
+);
+
+export default createRule<Options, MessageIds>({
+  name: 'no-invalid-this',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'disallow `this` keywords outside of classes or class-like objects',
+      category: 'Best Practices',
+      recommended: false,
+      extendsBaseRule: true,
+    },
+    messages: baseRule.meta.messages,
+    schema: [schema],
+  },
+  defaultOptions: [{ capIsConstructor: true }],
+  create(context) {
+    const rules = baseRule.create(context);
+    let argList: Array<string[]> = [];
+
+    return {
+      ...rules,
+      FunctionDeclaration(node: TSESTree.FunctionDeclaration): void {
+        const names = node?.params.map(
+          (param: TSESTree.Identifier) => param?.name,
+        );
+        argList.push(names);
```

Comment:
It would probably better to just store it as a `boolean[]` and do the work of checking for `this` up front.

```ts
function checkIfFunctionHasThisParam(node: Parameters): void {
  argsList.push(node.some(param =>
    param.type === AST_NODE_TYPES.Identifier && param.name === 'this',
  ));
}
```